ztools/iscsi/scan.go

```diff
@@ -0,0 +1,117 @@
+package iscsi
+
+import (
+	"net"
+	"fmt"
+	"regexp"
+)
+
+// FIXME, this is hacky and may not be very resilient
+var iscsiTargetRegex = regexp.MustCompile(`TargetName=([^\x00]+)[^T]*TargetAddress=([A-Za-z0-9\./:]+)`)
+/*var (
+	iscsiTargetNameRegex = regexp.MustCompile(`TargetName=([^\x00]+)`)
+	iscsiTargetAddressRegex = regexp.MustCompile(`TargetAddress=([A-Za-z0-9\./:]+)`)
+)*/
+
+
+func Scan(conn net.Conn, config *ISCSIConfig) (AuthLog, error) {
+	authlog := AuthLog{conn.RemoteAddr().String(), []Target{}, false}
+	// FIXME, there is too much hardcoded stuff going on here, needs more command line flags
+	body := fmt.Sprintf("%x00%x00%x00%x00%x00%x00%x00%x00%x00%x00%x",
+		"InitiatorName=iqn.1994-05.com.redhat:4fd2932635a0",
+		// Apparently login only accepts names that are % 4 == 1 bytes long.
+		// Integer division plus padding to the rescue!
+		"InitiatorAlias=" + fmt.Sprintf("%0" + fmt.Sprint(len(config.LocalLogin) / 4 * 4 + 5)  + "s", config.LocalLogin),
+		"SessionType=Discovery",
+		"HeaderDigest=None",
+		"DataDigest=None",
+		"DefaultTime2Wait=2",
+		"DefaultTime2Retain=0",
+		"IFMarker=No",
+		"OFMarker=No",
+		"ErrorRecoveryLevel=0",
+		"MaxRecvDataSegmentLength=32768",
+	)
+	header := fmt.Sprintf("438700000000%04x00023D00000000000000000000000000000000010000000000000000000000000000000000000000", len(body)/2)
+	data := []byte{}
+	fmt.Sscanf(header+body+"0000", "%x", &data)
+	conn.Write(data)
+	res := make([]byte, 1024)
+	_, err := conn.Read(res)
+	if err != nil {
+		authlog.HadError = true
+		return authlog, err
+	}
+
+	// fixme, "All" is hardcoded for now)
+	body = fmt.Sprintf("53656e64546172676574733d%x", "All")
+	all := fmt.Sprintf("048000000000%04x000000000000000000000001ffffffff000000010000000100000000000000000000000000000000%s00", (len(body) / 2) + 1, body)
+	data = []byte{}
+	fmt.Sscanf(all, "%x", &data)
+	conn.Write(data)
+	res = make([]byte, 1024)
+	_, err = conn.Read(res) //c.readUntilRegex(res, iscsiTargetNameRegex)
+	//c.grabData.Banner = string(res[0:n])
+	if err != nil {
+		authlog.HadError = true
+		return authlog, err
+	}
+	var haderror error
+	for _, bs := range iscsiTargetRegex.FindAllSubmatch(res, -1) {
```

Comment:
If we're making more connections, we need to set an upper bound on how many of these we process, so we don't DoS ourselves or the remote host.